test/e2e/e2e.go

```diff
@@ -1846,7 +1848,71 @@ var _ = ginkgo.Describe("e2e ingress traffic validation", func() {
 				}
 			}
 		})
+		// This test validates ingress traffic to nodeports with externalTrafficPolicy Set to local.
+		// It creates a nodeport service on both udp and tcp, and creates a backend pod on a single node.
+		// The backend pod is using the agnhost - netexec command which replies to commands
+		// with different protocols. We use the "hostname" command to have each backend pod to reply
+		// with its hostname.
+		// We use an external container to poke the service exposed on the node and ensure that only the
+		// nodeport on the node with the backend actually receives traffic.
+		// In case of dual stack enabled cluster, we iterate over all the nodes ips and try to hit the
+		// endpoints from both each node's ips.
+		// TODO: Add Test that verifies SNAT=GWRouterIP of ingress traffic does not occur with externalTrafficPolicy=local
+		ginkgo.It("Should be allowed to node local endpoints by nodeport services with externalTrafficPolicy=local", func() {
+			serviceName := "nodeportsvclocal"
+			ginkgo.By("Creating the nodeport service")
+			// Put endpoints only on one node 
+			localNode := nodes.Items[0]
+			framework.Logf("The only node with an endpoint is: %s", localNode.Name)
+			localEndpintsSelector := map[string]string{"hostname": localNode.Name}
+			npSpec := nodePortServiceSpecFrom(serviceName, endpointHTTPPort, endpointUDPPort, clusterHTTPPort, clusterUDPPort, localEndpintsSelector,v1.ServiceExternalTrafficPolicyTypeLocal)
+			np, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(context.Background(), npSpec, metav1.CreateOptions{})
+			nodeTCPPort, nodeUDPPort := nodePortsFromService(np)
+			framework.ExpectNoError(err)
+			ginkgo.By("Waiting for the single endpoint to pop up on node " + localNode.Name)
+			err = framework.WaitForServiceEndpointsNum(f.ClientSet, f.Namespace.Name, serviceName, 1, time.Second, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err, "failed to validate endpoints for service %s in namespace: %s", serviceName, f.Namespace.Name)
 
+			for _, protocol := range []string{"http", "udp"} {
+				for _, node := range nodes.Items {
+					for _, nodeAddress := range node.Status.Addresses {
+						// skipping hostnames
+						if !addressIsIP(nodeAddress) {
+							continue
+						}
+						responses := sets.NewString()
+						singleEndpointName := sets.NewString(localNode.Name + "-ep")
```

Comment:
I'd have an "expectedResponse" set here, that you fill with this or "TIMEOUT" depending on the node.
Then you always fill responses with what you get (either the pod's name OR timeout) and compare the two sets.